test/screenshot/run-screenshot-test.js

```diff
@@ -16,56 +16,194 @@
-        screenshotInfos.forEach(logScreenshotInfo);
-      },
-      (err) => {
-        console.error('\n\n\nERROR:\n\n');
-        console.error(err);
-      }
-    );
-}
+const SCREENSHOT_TEST_DIR_RELATIVE_PATH = 'test/screenshot/';
+const SCREENSHOT_TEST_DIR_ABSOLUTE_PATH = path.resolve(SCREENSHOT_TEST_DIR_RELATIVE_PATH);
 
-function handleUploadFailure(err) {
-  console.error('\n\nERROR uploading HTML/CSS/JS assets to Google Cloud Storage!\n\n');
-  console.error(err);
-}
+const readFileAsync = util.promisify(fs.readFile);
 
-function logScreenshotInfo(infoResponseBody) {
-  console.log(infoResponseBody.url);
+run();
 
-  // We don't use CBT's screenshot versioning features, so there should only ever be one version.
-  // Each "result" is an individual browser screenshot for a single URL.
-  infoResponseBody.versions[0].results.forEach((result) => {
-    logScreenshotInfoResult(result);
-  });
+async function run() {
+  const storage = new Storage();
 
-  console.log('');
-}
+  /**
+   * Unique timestamped directory path to prevent collisions between developers.
+   * @type {string}
+   */
+  const baseUploadDir = await storage.generateUniqueUploadDir();
+
+  /**
+   * @type {!Array<!UploadableTestCase>}
+   */
+  const testCases = [];
+
+  return uploadAllAssets();
+
+  async function uploadAllAssets() {
+    /**
+     * Relative paths of all asset files (HTML, CSS, JS) that will be uploaded.
+     * @type {!Array<string>}
+     */
+    const assetFileRelativePaths = glob.sync('**/*', {cwd: SCREENSHOT_TEST_DIR_ABSOLUTE_PATH, nodir: true});
+
+    return Promise.all(assetFileRelativePaths.map(uploadOneAsset))
+      .then(
+        () => logTestCases(),
+        (err) => Promise.reject(err)
+      );
+  }
+
+  /**
+   * @param {string} assetFileRelativePath
+   * @return {!Promise<!UploadableFile>}
+   */
+  async function uploadOneAsset(assetFileRelativePath) {
+    const assetFile = new UploadableFile({
+      destinationParentDirectory: `${baseUploadDir}assets/`,
+      destinationRelativeFilePath: assetFileRelativePath,
+      fileContent: await readFileAsync(`${SCREENSHOT_TEST_DIR_ABSOLUTE_PATH}/${assetFileRelativePath}`),
+    });
+
+    return storage.uploadFile(assetFile)
+      .then(
+        () => handleUploadOneAssetSuccess(assetFile),
+        (err) => handleUploadOneAssetFailure(err)
+      )
+      .then(
+        () => assetFile,
+        (err) => handleUploadOneAssetFailure(err)
+      );
+  }
+
+  /**
+   * @param {!UploadableFile} assetFile
+   * @return {!Promise<!Array<!UploadableFile>>}
+   */
+  async function handleUploadOneAssetSuccess(assetFile) {
+    const isHtmlFile = assetFile.destinationRelativeFilePath.endsWith('.html');
+    if (isHtmlFile) {
+      const testCase = new UploadableTestCase({htmlFile: assetFile});
+      testCases.push(testCase);
+      return capturePage(testCase);
+    }
+    return Promise.resolve([]);
+  }
+
+  /**
+   * @param {T} err
+   * @return {!Promise<T>}
+   * @template T
```

Comment:
Since this is a template, do we need to declare the type of the arg here?

https://github.com/material-components/material-components-web/blob/11b67cfcc48579fdffdd1f6840b3e123f65046da/test/screenshot/run-screenshot-test.js#L76-L84

Reply:
I don't _think_ so, but I could certainly be wrong.

OOC, how would you suggest declaring the type there?

I originally used `{*}` instead of `{T}`, but I wanted to make it clear that the type of the return value is the same as the argument.

WDYT?